lib/Lower/Bridge.cpp

```diff
@@ -574,41 +373,328 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+    if (info.stepExpr.has_value()) {
+      info.stepValue = genFIRLoopIndex(*info.stepExpr, type);
+    }
+    info.loopVariable = createTemporary(location, *info.loopVariableSym);
+
+    // Structured loop - generate fir.loop.
+    if (info.isStructured()) {
+      info.insertionPoint = builder->saveInsertionPoint();
+      llvm::SmallVector<mlir::Value, 1> stepVector;
+      if (info.stepValue) {
+        stepVector.push_back(info.stepValue);
+      }
+      info.doLoop = builder->create<fir::LoopOp>(location, lowerValue,
+                                                 upperValue, stepVector);
+      builder->setInsertionPointToStart(info.doLoop.getBody());
+      mlir::Value shadow = builder->create<fir::ConvertOp>(
+          location, info.loopVariableType, info.doLoop.getInductionVar());
+      localSymbols.pushShadowSymbol(*info.loopVariableSym, shadow);
+      return;
+    }
+
+    // Unstructured loop preheader code - initialize tripVariable, loopVariable.
+    if (!info.stepValue) {
+      info.stepValue = builder->create<mlir::ConstantOp>(
+          location, builder->getIntegerAttr(info.loopVariableType, 1));
+    }
+    mlir::Value tripCount =
+        builder->create<mlir::SubIOp>(location, upperValue, lowerValue);
+    tripCount =
+        builder->create<mlir::AddIOp>(location, tripCount, info.stepValue);
+    if (info.stepExpr.has_value()) {
+      tripCount = builder->create<mlir::SignedDivIOp>(location, tripCount,
+                                                      info.stepValue);
+    }
+    info.tripVariable = Fortran::lower::createTemporary(
+        location, *builder, localSymbols, info.loopVariableType, nullptr);
+    builder->create<fir::StoreOp>(location, tripCount, info.tripVariable);
+    builder->create<fir::StoreOp>(location, lowerValue, info.loopVariable);
+
+    // Unstructured loop header code - generate loop condition.
+    startBlock(info.headerBlock);
+    mlir::Value tripVariable =
+        builder->create<fir::LoadOp>(location, info.tripVariable);
+    mlir::Value zero = builder->create<mlir::ConstantOp>(
+        location, builder->getIntegerAttr(info.loopVariableType, 0));
+    mlir::Value cond = builder->create<mlir::CmpIOp>(
+        location, mlir::CmpIPredicate::sgt, tripVariable, zero);
+    genFIRConditionalBranch(cond, info.bodyBlock, info.successorBlock);
+  }
+
+  /// Generate FIR to end a structured or unstructured increment loop.
+  void genFIRIncrementLoopEnd(IncrementLoopInfo &info) {
+    mlir::Location location = toLocation();
+    if (info.isStructured()) {
+      // End fir.loop.
+      localSymbols.popShadowSymbol();
+      builder->restoreInsertionPoint(info.insertionPoint);
+      assert(info.doLoop.hasLastValue() && "loop must have a result");
+      mlir::Value result = builder->create<fir::ConvertOp>(
+          location, info.loopVariableType, info.doLoop.lastVal()[0]);
+      builder->create<fir::StoreOp>(location, result, info.loopVariable);
```

Comment:
OK. We don't (as far as I know) lower contained/inner procedures correctly AFAIK presently. (It's possible Jean implemented something and I've forgotten.) At any rate, host/free variables necessarily (in a theoretical sense) have to be (lambda) lifted and passed to the called procedure. By way of example, the above code becomes in pseudo-C:

```
int foo(int &a[10]) {
  for i = 1, 10 {
    store i -> i_mem;
    x = call bar(i_mem);
   store x -> a[i];
  }
}
int bar(int &i) {
  return i + 42;
}
```
So the lambda lifting is the promotion of `i` into an argument. (There are various implementation approaches.  We're just concerned about the semantics.)

The second problem is that the value `i` isn't immediately defined in a way to be used as pass-by-reference. For Fortran 77 support, we'll want to lower this value into a temporary memory location to be passed by reference to the called procedure. This should be the default for all calls that we arrive at that have (ssa) values and not memory references (that we can load from).

The corollary question is what happens if `bar` changes the value in the temporary `i_mem`. I believe that is disallowed by the standard, as a `DO` loop shouldn't modify the LCV.

Reply:
My example is a bit narrowing. The issue is more general. The lambda case is the uglier case, but the general issue is that I do not find any Fortran rule that prevent a `do-variable` to be associated (through whatever association process you can imagine, there may be some that are impossible, though). In f90, you can (I think) also have a pointer pointing at it.

Another example where I cannot find why it would be forbidden:
```
subroutine foo2(a)
  integer :: i, a(10)
  target :: i
  integer, pointer :: p
  p => i
  do i=1, 10
    a(i) = p
  end do
end subroutine
```
Both gfortran and pgfortran compiles these two example just fine and update a(i) based on the value of `i` inside the loop as expected.

So my point is more to say that until a deeper analysis of the `do-variable` associations, it is unsafe to not store it at every loop iteration (lambda or not lambda).